### PR TITLE
schema: add KeyPoint3D, KeyLine3D, Polygon3D

### PR DESCRIPTION
**Pull request by nehalmamgain**
- add 3D equivalents in Cartesian coordinates in sensor space for KeyPoint2D,
KeyLine2D and Polygon2D on the image plane respectively.

This PR is based on Nisse's PR https://github.com/TRI-ML/dgp/pull/45
I don't have write access to the aformentioned PR so I opened a new PR with the changes asked by Quincy, after clarifications from Nisse.

Sorry for the incovenience @tk-woven ; let's continue discussion on this  PR.

---
This change is [Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101)

### CODE REVIEW

**Comment by nehalmamgain** [comment]
@tk-woven 

> Could you please share where we can find documentation of keys and value semantics to expect here?

You can see some of the `attributes` [here](https://docs.google.com/document/d/1KsbyvF7KMIDNZSQFTsS5rYO-9-Jz0wIlm1Jg462BA0I/edit#bookmark=id.jgrjixh7famn). Please note that the contents for `attributes` differ based on the project - linked ones are for lane lines. I do not know which values for the other projects.

**Comment by nehalmamgain** [comment]
Btw please let me know if you somehow don't have access to the doc (it is in TME drive).

**Review by tk-woven — COMMENTED** [review]
+reviewer:@kuanleetri 

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8Mj8s922GgfELzF_-1:b-66yv47)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @kuanleetri)

**Review by tk-woven — CHANGES_REQUESTED** [review]
Thanks for the link to the doc. I am able to view it!

Let's add some usage instructions for the attributes map in the code. I went into detail in a company-internal Slack exchange about potential issues with that representation which I won't repeat in full here. Adding "usage" docs to the attributes map is a fair middle ground. Basically, we want to discourage people from putting arbitrary data in that map, so the middle ground would be instructing users to only add key/value pairs that are stored in some document accessible by all potential users of the data in their project. This way people parsing the data have a guide on how to interpret values, for example when they need to cast or do bounds-checking.

This doesn't stop anyone from violating the semantics, which is why I call it a middle ground.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8MkJ8_D_QP2u3usZFg:b-jmb6ti)* status: all files reviewed, 3 unresolved discussions (waiting on @kuanleetri and @nehalmamgain)

___
*[`dgp/proto/annotations.proto` line 251 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjN9S8ldTqL-rCmlg:-N8MjN9S8ldTqL-rCmlh:bw6zj3p) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L251)):*
> ```protobuf
> // 3D point annotation.
> message KeyPoint3DAnnotation {
>   // Class identifier (should be in [0, num_classes - 1])
> ```

Can you share where we can find the `num_classes` value please?
___
*[`dgp/proto/annotations.proto` line 258 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjnCH6HsM10lhDbQk:-N8MjnCH6HsM10lhDbQl:b-poi02t) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L258)):*
> ```protobuf
> 
>   // An associative map stores arbitrary attributes, where the key is attribute name
>   // and the value is attribute value. Both key_type and value_type are string.
> ```

Let's condense this to

```
// A map of attribute names to their values.
```
___
*[`dgp/proto/annotations.proto` line 261 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjwEC0Rtb2vtHrcZ9:-N8MjwEDCXqfLzCZjENW:bvi8nv6) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L261)):*
> ```protobuf
>   map<string, string> attributes = 3;
> 
>   // An identifier key. Used to link with other annotations.
> ```

How does one establish a link with this string?

**Review by nehalmamgain — COMMENTED** [review]
Understood, thanks! Will make the change

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8MpDCV5hNUbGE17Qo5:b-gap235)* status: all files reviewed, 3 unresolved discussions (waiting on @kuanleetri, @pd-nisse, and @tk-woven)

___
*[`dgp/proto/annotations.proto` line 251 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjN9S8ldTqL-rCmlg:-N8MpNX78XMnozJ_Yf7R:bgn3xbr) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L251)):*
<details><summary><i>Previously, tk-woven wrote…</i></summary><blockquote>

Can you share where we can find the `num_classes` value please?
</blockquote></details>

I brought this comment over from the previous PR, but basically this value (total number of classes) will differ ontology to ontology.
It's hard to say `num_classes` is a fixed value. 
You would like me to add some comment explaining what's expected in `num_classes`?
___
*[`dgp/proto/annotations.proto` line 258 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjnCH6HsM10lhDbQk:-N8MpqMj99WHLvRxxY3L:btn5jp1) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L258)):*
<details><summary><i>Previously, tk-woven wrote…</i></summary><blockquote>

Let's condense this to

```
// A map of attribute names to their values.
```
</blockquote></details>

Gotcha
___
*[`dgp/proto/annotations.proto` line 261 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjwEC0Rtb2vtHrcZ9:-N8Mpu_WD0u-3IeUEici:b6yqays) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L261)):*
<details><summary><i>Previously, tk-woven wrote…</i></summary><blockquote>

How does one establish a link with this string?
</blockquote></details>

I can't seem to, but could you please tag @pd-nisse for this question?

**Review by tk-woven — CHANGES_REQUESTED** [review]
Thank you!

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8Mqku57UmNBCZ1xFM2:b-9755fn)* status: all files reviewed, 3 unresolved discussions (waiting on @kuanleetri, @nehalmamgain, @pd-nisse, and @tk-woven)

___
*[`dgp/proto/annotations.proto` line 251 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjN9S8ldTqL-rCmlg:-N8MqN633n_bJ_hdWK_z:b-jt7tjj) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L251)):*
<details><summary><i>Previously, nehalmamgain wrote…</i></summary><blockquote>

I brought this comment over from the previous PR, but basically this value (total number of classes) will differ ontology to ontology.
It's hard to say `num_classes` is a fixed value. 
You would like me to add some comment explaining what's expected in `num_classes`?
</blockquote></details>

Thanks. The way it's styled makes me think we can find `num_classes` in some message definition or other code. If it differs based on the ontology, let's please note that in the comment, basically something like

```
// Class identifier (should be in [0, num_classes_in_your_ontology -1]).
```
___
*[`dgp/proto/annotations.proto` line 261 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjwEC0Rtb2vtHrcZ9:-N8MqgIf3YPBWI-Vz7sb:b-5tpdgf) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L261)):*
<details><summary><i>Previously, nehalmamgain wrote…</i></summary><blockquote>

I can't seem to, but could you please tag @pd-nisse for this question?
</blockquote></details>

@pd-nisse , could you let us know, please?

**Review by nehalmamgain — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101)* status: all files reviewed, 3 unresolved discussions (waiting on @kuanleetri, @nehalmamgain, @pd-nisse, and @tk-woven)

___
*[`dgp/proto/annotations.proto` line 261 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjwEC0Rtb2vtHrcZ9:-N8MwAve326L1PxJRaJB:b-9hnv8a) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L261)):*
<details><summary><i>Previously, tk-woven wrote…</i></summary><blockquote>

@pd-nisse , could you let us know, please?
</blockquote></details>

P.S. While editing the file to factor in your suggestions, I believe this section is just copied over from KeyPoint2D
```
  // An identifier key. Used to link with other annotations.
  string key = 4;
```

But yes, let's wait for Nisse to provide more context here.

**Review by nehalmamgain — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101)* status: 1 of 2 files reviewed, 3 unresolved discussions (waiting on @kuanleetri, @pd-nisse, and @tk-woven)

___
*[`dgp/proto/annotations.proto` line 251 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjN9S8ldTqL-rCmlg:-N8Myyaf9RRn3lOYlqAr:b-896fix) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L251)):*
<details><summary><i>Previously, tk-woven wrote…</i></summary><blockquote>

Thanks. The way it's styled makes me think we can find `num_classes` in some message definition or other code. If it differs based on the ontology, let's please note that in the comment, basically something like

```
// Class identifier (should be in [0, num_classes_in_your_ontology -1]).
```
</blockquote></details>

Done.
___
*[`dgp/proto/annotations.proto` line 258 at r1](https://reviewable.io/reviews/tri-ml/dgp/101#-N8MjnCH6HsM10lhDbQk:-N8MyzrH0fJ6VACvvdi1:b-896fix) ([raw file](https://github.com/tri-ml/dgp/blob/6c0367742c3c83df51250a79d12048d4a9bbbe24/dgp/proto/annotations.proto#L258)):*
<details><summary><i>Previously, nehalmamgain wrote…</i></summary><blockquote>

Gotcha
</blockquote></details>

Done.

**Review by tk-woven — COMMENTED** [review]
Thanks, Nehal! Waiting on Nisse's feedback, but otherwise looks good to me :)

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8N-X5-CVUAR31PJmZD:b-3z2c6y)* status: all files reviewed, 1 unresolved discussion (waiting on @kuanleetri and @pd-nisse)

**Comment by nehalmamgain** [comment]
Cool, thanks!

**Comment by pd-nisse** [comment]
@nehalmamgain @tk-woven : You are correct, the `key` attribute was simply copied from `KeyPoint2DAnnotation`, to have an analogous solution.
Since Points/LineAnnotations do not have an `instance_id`, the attribute `key` is used to cross-reference them. An example would be different skeleton points. Those can then be referenced on the parent "Pedestrian" `BoundingBox2DAnnotation` as relating to that specific instance.

**Review by tk-woven — COMMENTED** [review]
Thanks Nisse :) I'd like us to add some documentation to the `key` field about how to determine appropriate values to set the string to. Nehal, are you familiar with the work enough to extend the docs with information on selecting proper values, or do we need further explanation?

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8V_wJXC5S7h97pjHkO:b3fo0b5)* status: all files reviewed, 1 unresolved discussion (waiting on @kuanleetri and @pd-nisse)

**Review by nehalmamgain — COMMENTED** [review]
Thanks for the explanation Nisse!

> Nehal, are you familiar with the work enough to extend the docs with information on selecting proper values, or do we need further explanation?

I understood what Nisse meant. Having never worked with `KeyPoint2D` (and not seen the 3D jsons that will be produced after this PR merges), I'm not sure of the `proper values` but I can look it up. 
Since you've worked with the OD DGP datasets before (and they do contain key point 2D information), if you would like to extend the docs yourself, please let me know! 

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8W9VWU0Vol979KgJgV:b-3ze37q)* status: all files reviewed, 1 unresolved discussion (waiting on @kuanleetri and @pd-nisse)

**Review by tk-woven — COMMENTED** [review]
Thanks! Personally, I still have very little idea of what valid values of `key` could be or how to use that field, so I defer to the authors here and people that have put that key to use before.

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8WC1XHAXEmaY4k336J:b8a1vqj)* status: all files reviewed, 1 unresolved discussion (waiting on @kuanleetri and @pd-nisse)

**Review by nehalmamgain — COMMENTED** [review]
> I still have very little idea of what valid values of key could be or how to use that field, 

Me either, I have never used this field.
Let me just investigate an OD dataset since this is still blocking the PR.

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8Wp4HW6BMy8Gq6NSkb:b2hbjkv)* status: all files reviewed, 1 unresolved discussion (waiting on @kuanleetri and @pd-nisse)

**Review by nehalmamgain — COMMENTED** [review]
@tk-woven There seems to be a string value like `"key_point_2d_key": "0fe8ec63-adce-4737-a8db-d39e918c9ce0"` for a bounding box instance like a car, and associated key point json has the corresponding `"key": "0fe8ec63-adce-4737-a8db-d39e918c9ce0"`. How do you wish to proceed?

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8WpuBA4oN-oiA7_1Da:btrmgyw)* status: all files reviewed, 1 unresolved discussion (waiting on @kuanleetri and @pd-nisse)

**Review by tk-woven — COMMENTED** [review]
Thanks Nehal. So, it seems that all annotations related to some ground truth instance associate themselves with the instance by using the instance's key as a value for their own `key` field. Let's update the docs to note this, and then I believe we're good to go :)

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8WyNzJ1Z9BbiAtMewA:b21bqa2)* status: all files reviewed, 1 unresolved discussion (waiting on @kuanleetri and @pd-nisse)

**Review by tk-woven — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @kuanleetri)

**Review by tk-woven — APPROVED** [review]
:lgtm: , but it looks like we might have some CI failures unfortunately.

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8XZ30y9vJsBMjunTMm:bszfqtq)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @kuanleetri)

**Comment by nehalmamgain** [comment]
Force pushing trivial change (and undoing it) to restart the CI
Basically
```
.pylintrc:1: [E0015(unrecognized-option), ] Unrecognized option found: accept-no-param-doc, accept-no-return-doc, accept-no-yields-doc
``` 
should not be showing up for a `*_pb2.py` file which is in ignore list.

**Comment by nehalmamgain** [comment]
Still fails...
@tk-woven I will look to your support for merging this please 🙇‍♀️

**Review by kuanleetri — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r1, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8_dYDFBDpA3kCjp1Ru:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nehalmamgain)

**Review by kuanleetri — COMMENTED** [review]
Pass the test should be good to go.

> *[Reviewable](https://reviewable.io/reviews/tri-ml/dgp/101#-:-N8_ddNT7TASd7OQUFtm:b-ho948x)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nehalmamgain)

**Comment by tk-woven** [comment]
Thanks Kuan! Nehal, I'll look into the CI issue today.

**Comment by tk-woven** [comment]
The problem in CI is that the `check_docs` pylint plugin is not available, and so the configuration we're providing is invalid. Looking into how to fix this.

**Comment by tk-woven** [comment]
Fix for CI issue is in #109. It's the only issue I have identified at this time.

**Comment by nehalmamgain** [comment]
> Pass the test should be good to go.

Thanks @kuanleetri 
Although this PR was [passing](https://github.com/TRI-ML/dgp/actions/runs/2772613990) the CI test before, it fails out of the blue now. Tyler kindly posted a [fix](https://github.com/TRI-ML/dgp/pull/109) as mentioned. Please review when convenient 🙇‍♀️

**Comment by tk-woven** [comment]
I'm guessing we picked up a new pylint version with a recent Docker build on master. We could confirm if anyone has an older image laying around, but I don't think it's mandatory since the fix will take care of the issue.

**Comment by tk-woven** [comment]
#109 is merged, so let's give CI another shot here!

**Comment by nehalmamgain** [comment]
Thanks again Tyler 🙂

**Pull request merged**
